private/buf/buflsp/buflsp.go

```diff
@@ -0,0 +1,269 @@
+
+	// TODO: set up logging. We need to forward everything from server.logger through to
+	// the client, if tracing is turned on. The right way to do this is with an extra
+	// goroutine and some channels.
+
+	return nil
+}
+
+// findImportable finds all files that can potentially be imported by the proto file at
+// uri. This returns a map from potential Protobuf import path to the URI of the file it would import.
+//
+// Note that this performs no validation on these files, because those files might be open in the
+// editor and might contain invalid syntax at the moment. We only want to get their paths and nothing
+// more.
+func (l *lsp) findImportable(
+	ctx context.Context,
+	uri protocol.URI,
+) (map[string]bufimage.ImageFileInfo, error) {
+	fileInfos, err := l.controller.GetImportableImageFileInfos(ctx, uri.Filename())
+	if err != nil {
+		return nil, err
+	}
+
+	imports := make(map[string]bufimage.ImageFileInfo)
+	for _, fileInfo := range fileInfos {
+		imports[fileInfo.Path()] = fileInfo
+	}
+
+	l.logger.Sugar().Debugf("found imports for %q: %#v", uri, imports)
+
+	return imports, nil
+}
+
+// newHandler constructs an RPC handler that wraps the default one from jsonrpc2. This allows us
+// to inject debug logging, tracing, and timeouts to requests.
+func (l *lsp) newHandler() jsonrpc2.Handler {
+	actual := protocol.ServerHandler(newServer(l), nil)
+	return func(ctx context.Context, reply jsonrpc2.Replier, req jsonrpc2.Request) (retErr error) {
+		ctx, span := l.tracer.Start(
+			ctx,
+			tracing.WithErr(&retErr),
+			tracing.WithAttributes(attribute.String("method", req.Method())),
+		)
+		defer span.End()
+
+		l.logger.Debug(
+			"processing request",
+			zap.String("method", req.Method()),
+			zap.ByteString("params", req.Params()),
+		)
+
+		ctx = withRequestID(ctx)
+
+		replier := l.adaptReplier(reply, req)
+
+		// Verify that the server has been initialized if this isn't the initialization
+		// request.
+		if req.Method() != protocol.MethodInitialize && l.initParams.Load() == nil {
+			return replier(ctx, nil, fmt.Errorf("the first call to the server must be the %q method", protocol.MethodInitialize))
+		}
+
+		return actual(ctx, replier, req)
+	}
+}
+
+// adaptReplier wraps a jsonrpc2.Replier, allowing us to inject logging and tracing and so on.
+func (l *lsp) adaptReplier(reply jsonrpc2.Replier, req jsonrpc2.Request) jsonrpc2.Replier {
+	return func(ctx context.Context, result any, err error) error {
+		if err != nil {
+			l.logger.Warn(
+				"responding with error",
+				zap.String("method", req.Method()),
+				zap.Error(err),
+			)
+		} else {
+			l.logger.Debug(
+				"responding",
+				zap.String("method", req.Method()),
+				zap.Reflect("params", result),
+			)
+		}
+
+		return reply(ctx, result, err)
+	}
+}
+
+// connAdapter wraps a connection and logs calls and notifications.
+//
+// By default, the ClientDispatcher does not log the bodies of requests and responses, making
+// for much lower-quality debugging.
```

Comment:
Even though this is pretty small, I think it's reasonable to split this out in a different file, since this is fairly isolated functionality (even though I see the reasoning to keep it side-by-side with `adaptReplier`/`wrapReplier`).